### PR TITLE
Implement RFC 1007 - Frozen users for parsec v2

### LINKED ISSUE #5811: API User freeze/unfreeze
Add server administration route to freeze and unfreeze users.

Frozen users won't be able to connect the parsec servers.

This allows for automated processes to temporarily block users while waiting for an organization administrator to actually (and cryptographically) revoke the corresponding user

### PR DESCRIPTION
**Pull request by vxgmichel**
Close #5811
Related #5817

Remaining tasks
- [x] Add news fragment
- [x] Disconnect client when frozen
- [x] Add GUI error message
- [x] Add server administration documentation

### CODE REVIEW

File: parsec/backend/asgi/administration.py [hunk 1]
```diff
@@ -285,3 +294,98 @@ async def administration_server_stats() -> Response:
...
+            user = await backend.user.get_user_from_email(organization_id, user_email)
+        elif user_id is not None:
+            user = await backend.user.get_user(organization_id, user_id)
+        else:
+            assert False
+    except UserNotFoundError:
+        await user_not_found_abort()
+
+    # Make sure the user is valid
+    if user.human_handle is None or user.revoked_on is not None:
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/backend/asgi/administration.py:378]: 549c5e1ae

File: parsec/backend/user.py [hunk 2]
```diff
@@ -343,6 +355,25 @@ async def get_user_with_device(
         """
         raise NotImplementedError()
 
+    async def get_user_from_email(self, organization_id: OrganizationID, user_email: str) -> User:
+        """
```
Comment (mmmarcos) [inline parsec/backend/user.py:359]: I would add some info regarding the email vs parsec id discussion: 
```suggestion
        """
        Returns the (non-revoked) user associated with the given email address.
```

**Review by mmmarcos — COMMENTED** [review]
LGTM, just a couple of comments.

Also: we should add a documentation somewhere in the admin guide as this route is meant to be used externally (probably via script, webhook or cron job...).

File: parsec/backend/postgresql/migrations/datamodel.sql [hunk 3]
```diff
@@ -151,7 +151,7 @@ CREATE TABLE shamir_recovery_share (
 
 
 -- Alter user table to introduce a cross-reference between user id and shamir id
-ALTER TABLE user_ ADD shamir_recovery INTEGER REFERENCES shamir_recovery_setup (_id);
+ALTER TABLE user_ ADD FOREIGN KEY (shamir_recovery) REFERENCES shamir_recovery_setup (_id);
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/backend/postgresql/migrations/datamodel.sql:154]: No impact, this is done to keep the right column ordering in the `user_` table

File: parsec/backend/user.py (hunk 2 above)
Comment (vxgmichel) [reply parsec/backend/user.py:359]: 1c9f1322aeaabe5880b1c6abdc5ecb928db70cc1

File: parsec/backend/user.py [hunk 4]
```diff
@@ -343,6 +355,25 @@ async def get_user_with_device(
...
+        Raises:
+            UserNotFoundError
+        """
+        results, _ = await self.find_humans(
+            organization_id=organization_id,
+            query=user_email,
+            omit_revoked=True,
+            omit_non_human=True,
+        )
+        for item in results:
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/backend/user.py:375]: 1c9f1322aeaabe5880b1c6abdc5ecb928db70cc1

File: docs/adminguide/frozen_users.rst [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+.. _doc_adminguide_frozen_users:
+
+
+Frozen users
+============
+
+It is possible for server administrators to freeze specific Parsec users.
+
+This feature is not part of the Parsec revocation system, but a way for system administrators to temporarily block users from connecting to the server while waiting for organization administrators to actually revoke them. It is exposed as an HTTP route that requires the administration token. This way, it is possible for existing directory services to automatically freeze users that have been removed from the directory.
```
Comment (mmmarcos) [inline docs/adminguide/frozen_users.rst:11]: I would rephrase it a little bit:
* add small note of what revocation system is about.
* avoid using "feature" as it is a more spec-related term.
* emphasize main goal of the freeze/unfreeze mechanism 
```suggestion
Frozen users will be temporarily blocked from connecting to the Parsec server while waiting for an organization administrator to actually revoke them. This operation is thus not part of the Parsec revocation system (i.e. the definitive removal of the user's rights).

The freeze mechanism enables to automatically block users who have been deleted from a directory service (such as OpenLDAP or AD), as it is exposed in the form of HTTP routes that only require an administration token. 
```

File: docs/adminguide/frozen_users.rst [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  .. code-block:: json
+
+    {
+      "frozen": true,
+      "user_email": "alice@example.com",
+      "user_id": "940a380aedd44127863d952a66cfce1e",
+      "user_name": "Alice"
+    }
+
+User identification: email address vs Parsec ID
```
Comment (mmmarcos) [inline docs/adminguide/frozen_users.rst:96]: I would move this to the top section (before `HTTP users route`).
Also, I would rename it `Notes on user identification` or just `User identification` and add a label to reference it.

File: docs/adminguide/frozen_users.rst [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+It is possible for server administrators to freeze specific Parsec users.
+
+This feature is not part of the Parsec revocation system, but a way for system administrators to temporarily block users from connecting to the server while waiting for organization administrators to actually revoke them. It is exposed as an HTTP route that requires the administration token. This way, it is possible for existing directory services to automatically freeze users that have been removed from the directory.
+
+HTTP ``users`` route
+--------------------
+
+This route is made available as ``/administration/organizations/<raw_organization_id>/users`` and requires an administration token.
+
+It only supports the ``GET`` method which lists information for all users, including Parsec ID, user name, user email, and frozen status.
```
Comment (mmmarcos) [inline docs/adminguide/frozen_users.rst:18]: I'd use a list here:
```suggestion
It only supports the ``GET`` method which lists information for all users, including:
* Parsec ID
* user name
* user email
* frozen status
```

File: docs/adminguide/frozen_users.rst [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+Here's an example using ``curl`` and ``jq``:
+
+  .. code-block:: bash
+
+    $ curl https://$server/administration/organizations/$organization/users \
+        -H "Authorization: Bearer $administration_token" | jq
+
+A successful request returns a JSON dictionary similar to the one below:
```
Comment (mmmarcos) [inline docs/adminguide/frozen_users.rst:27]: I think it's _object_ instead of _dictionary_ in JSON terminology.
```suggestion
A successful request returns a JSON object with the following structure:
```

File: docs/adminguide/frozen_users.rst [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        {
+          "user_name": "Bob",
+          "frozen": false,
+          "user_email": "bob@example.com",
+          "user_id": "7d04b0df51a74158b35b6386eecdf4e0"
+        }
+      ]
+    }
+
+HTTP ``users/frozen`` route
```
Comment (mmmarcos) [inline docs/adminguide/frozen_users.rst:48]: Ouch, I didn't see it before: shouldn't be `users/freeze`? it is a `POST` action

File: docs/adminguide/frozen_users.rst [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+---------------------------
+
+This route is made available as ``/administration/organizations/<raw_organization_id>/users/frozen`` and requires an administration token.
+
+It only supports the ``POST`` method which modifies the ``frozen`` status for a given user.
+
+Here's an example of generating the request data using ``jq``:
```
Comment (mmmarcos) [inline docs/adminguide/frozen_users.rst:55]: It seems odd to use `curl` for the get example and `jq` for the post example, maybe we should stick to one tool? (not 100% sure of my comment though)

File: docs/adminguide/frozen_users.rst [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+  .. code-block:: bash
+
+    $ DATA=$(jq -n --arg user_id "$user_id" --argjson frozen true '$ARGS.named')
+    $ echo $DATA
+    {
+    "user_id": "940a380aedd44127863d952a66cfce1e",
+    "frozen": true
+    }
+
+The request can also use the ``user_email`` field instead of ``user_id`` to identify the Parsec user:
```
Comment (mmmarcos) [inline docs/adminguide/frozen_users.rst:66]: We could add a link to the section (see comment below):
```suggestion
The request can also use the ``user_email`` field instead of ``user_id`` to identify the Parsec user (see :ref:`user-identification`):
```

File: docs/adminguide/frozen_users.rst (hunk 6 above)
Comment (mmmarcos) [inline docs/adminguide/frozen_users.rst:96]: I would move this to the top section (before `HTTP users route`).
Also, I would rename it `Notes on user identification` or just `User identification` and add a label to reference it.

```suggestion
.. _user-identification:

User identification
```

File: docs/adminguide/frozen_users.rst [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+-----------------------------------------------
+
+There is a subtle difference between the two ways to identify a user. At any given time, an email address can be used to uniquely identify a non-revoked user from a given organization. In contrast, a Parsec user ID identifies uniquely any user from all organizations in the Parsec server, including revoked users. This means that over time, an email address can identify different Parsec users with different Parsec IDs, even from the same organization.
+
+The frozen status configured by the ``POST`` method is specifically associated with the Parsec user ID, regardless of the identification method used in the request body. This has the following consequence: if a user is revoked and then a new user is created with the same email address, the frozen status will **not** be applied to the new user.
+
+Error handling
+--------------
+
+The following standard errors are handled the same way as for the other administration routes:
```
Comment (mmmarcos) [inline docs/adminguide/frozen_users.rst:106]:
```suggestion
The following standard errors are handled the same way as for other administration routes:
```

Comment (mmmarcos) [reply docs/adminguide/frozen_users.rst:106]: Actually it's weird to talk about "other administration routes" since (i think) this is the very first time we publicly document an administration route.

**Review by mmmarcos — COMMENTED** [review]
LGTM. I suggest some changes to the docs.

IMO the `users/frozen` route should be named `users/freeze`. I didn't see that before (sorry!) and I don't remembrer if we've discussed it.

File: docs/adminguide/frozen_users.rst (hunk 10 above)
Comment (vxgmichel) [reply docs/adminguide/frozen_users.rst:55]: Both examples uses `curl` to perform the request and `jq` to generate or prettify the json data.

File: docs/adminguide/frozen_users.rst [hunk 13]
```diff
@@ -0,0 +1,114 @@
+.. Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
+.. _doc_adminguide_frozen_users:
+
+
+Frozen users
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply docs/adminguide/frozen_users.rst]: [90cb593](https://github.com/Scille/parsec-cloud/pull/5842/commits/90cb593b227fb2624aa500b10763bcd7427c667c)

File: docs/adminguide/frozen_users.rst (hunk 9 above)
Comment (vxgmichel) [reply docs/adminguide/frozen_users.rst]: [76af22d](https://github.com/Scille/parsec-cloud/pull/5842/commits/76af22d60059ac1f670d9167ada9f0bcb3f70d2c)

File: docs/adminguide/frozen_users.rst (hunk 12 above)
Comment (vxgmichel) [reply docs/adminguide/frozen_users.rst]: [70d6ef2](https://github.com/Scille/parsec-cloud/pull/5842/commits/70d6ef29491522ed83c28a1dcd9f084b78896ef2)

File: docs/adminguide/frozen_users.rst (hunk 6 above)
Comment (vxgmichel) [reply docs/adminguide/frozen_users.rst]: [70d6ef2](https://github.com/Scille/parsec-cloud/pull/5842/commits/70d6ef29491522ed83c28a1dcd9f084b78896ef2)

File: docs/adminguide/frozen_users.rst (hunk 8 above)
Comment (vxgmichel) [reply docs/adminguide/frozen_users.rst]: [70d6ef2](https://github.com/Scille/parsec-cloud/pull/5842/commits/70d6ef29491522ed83c28a1dcd9f084b78896ef2)

File: docs/adminguide/frozen_users.rst (hunk 7 above)
Comment (vxgmichel) [reply docs/adminguide/frozen_users.rst]: [70d6ef2](https://github.com/Scille/parsec-cloud/pull/5842/commits/70d6ef29491522ed83c28a1dcd9f084b78896ef2)

File: docs/adminguide/frozen_users.rst (hunk 6 above)
Comment (vxgmichel) [reply docs/adminguide/frozen_users.rst]: [70d6ef2](https://github.com/Scille/parsec-cloud/pull/5842/commits/70d6ef29491522ed83c28a1dcd9f084b78896ef2)

File: docs/adminguide/frozen_users.rst (hunk 5 above)
Comment (vxgmichel) [reply docs/adminguide/frozen_users.rst]: [70d6ef2](https://github.com/Scille/parsec-cloud/pull/5842/commits/70d6ef29491522ed83c28a1dcd9f084b78896ef2)

File: docs/adminguide/frozen_users.rst (hunk 10 above)
Comment (mmmarcos) [reply docs/adminguide/frozen_users.rst]: oops, my bad

**Review by mmmarcos — APPROVED** [review]
LGTM
